qualtran/surface_code/azure_cost_model.py

```diff
@@ -13,55 +13,75 @@
-from qualtran.surface_code.data_block import FastDataBlock
-from qualtran.surface_code.quantum_error_correction_scheme_summary import (
-    QuantumErrorCorrectionSchemeSummary,
-)
-from qualtran.surface_code.rotation_cost_model import RotationCostModel
+import attrs
+
+from qualtran.resource_counting import GateCounts
+
+if TYPE_CHECKING:
+    from qualtran.surface_code import (
+        AlgorithmSummary,
+        QuantumErrorCorrectionSchemeSummary,
+        RotationCostModel,
+    )
 
 
 def minimum_time_steps(
-    error_budget: float, alg: AlgorithmSummary, rotation_model: RotationCostModel
+    *, error_budget: float, alg: 'AlgorithmSummary', rotation_model: 'RotationCostModel'
 ) -> int:
     r"""Minimum number of time steps needed for the algorithm.
 
     Equals
     $$
         M_\mathrm{meas} + M_R + M_T + 3 M_\mathrm{Tof} + D_R \textrm{rotation cost}
     $$
+
     Where:
-        $M_\mathrm{meas}$ is the number of measurements.
-        $M_R$ is the number of rotations.
-        $M_T$ is the number of T operations.
-        $M_mathrm{Tof}$ is the number of toffoli operations.
-        $D_R$ is the depth of the rotation circuit.
-        $\textrm{rotation cost}$ is the number of T operations needed to approximate a rotation to $\epsilon/(3*M_R)$.
-    Source: Equation D3 in https://arxiv.org/abs/2211.07629.
+     - $M_\mathrm{meas}$ is the number of measurements.
+     - $M_R$ is the number of rotations.
+     - $M_T$ is the number of T operations.
+     - $M_mathrm{Tof}$ is the number of toffoli operations.
+     - $D_R$ is the number of layers containing at least one rotation. This can be smaller than
+       the total number of non-Clifford layers since it excludes layers consisting only of T or
+       Toffoli gates.
+     - $\textrm{rotation cost}$ is the number of T operations needed to approximate a rotation to $\epsilon/(3*M_R)$.
+
+    Reference:
+        https://arxiv.org/abs/2211.07629.
+        Equation D3.
 
     Args:
-        error_budget: Error Budget.
-        alg: A summary of an algorithm/circuit.
+        error_budget: The total error budget. One third is prescribed to be used for rotation
+            synthesis.
+        alg: The logical algorithm costs, from which we extract the gate count.
         rotation_model: Cost model used to compute the number of T gates
             needed to approximate rotations.
     """
-    c_min = math.ceil(alg.measurements + alg.rotation_gates + alg.t_gates + 3 * alg.toffoli_gates)
+    M = alg.n_logical_gates.total_beverland_count()
+    c_min = M['meas'] + M['R'] + M['T'] + 3 * M['Tof']
     eps_syn = error_budget / 3
-    if alg.rotation_gates > 0:
-        rotation_cost = rotation_model.rotation_cost(eps_syn / alg.rotation_gates)
-        c_min += math.ceil(
-            alg.rotation_circuit_depth * (rotation_cost.n_t + 4 * rotation_cost.n_ccz)
-        )
+    if M['R'] > 0:
+        # NOTE! argument to the rotation_cost method is inverted relative to eq. D3.
```

Comment:
Thanks, yeah the mathematics comes out the same but if you're following along it's important to call out notational differences especially since the negative sign isn't apparent in this code snippet. I'll update the comment